From 17e0db31c6b31d66730afc1d9eb20ba65862f120 Mon Sep 17 00:00:00 2001
From: Ilkka Koskinen <ilkka.koskinen@intel.com>
Date: Mon, 22 Sep 2014 10:45:39 -0700
Subject: [PATCH 098/117] atomisp: Switch of ISP if firmware loading fails

ISP is always enabled at boot up. If ISP firmware load fails, ISP
needs  to be switched off. Otherwise S0ix states remain blocked.

Change-Id: Ibcf766cd48514c7d0994e7819321c39b0fb9de67
Signed-off-by: Ilkka Koskinen <ilkka.koskinen@intel.com>
---
 .../pci/atomisp2/atomisp_driver/atomisp_v4l2.c     | 24 ++++++++++++++++++++++
 1 file changed, 24 insertions(+)

diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_v4l2.c b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_v4l2.c
index 57049d6..4761e84 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_v4l2.c
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_v4l2.c
@@ -1140,6 +1140,7 @@ static int atomisp_pci_probe(struct pci_dev *dev,
 	unsigned int start;
 	void __iomem *base;
 	int err;
+	u32 irq;
 
 	if (!dev) {
 		dev_err(&dev->dev, "atomisp: error device ptr\n");
@@ -1431,6 +1432,29 @@ wdt_work_queue_fail:
 fw_validation_fail:
 	release_firmware(isp->firmware);
 load_fw_fail:
+	/*
+	 * Switch off ISP, as keeping it powered on would prevent
+	 * reaching S0ix states.
+	 *
+	 * The following lines have been copied from atomisp suspend path
+	 */
+
+	pci_read_config_dword(dev, PCI_INTERRUPT_CTRL, &irq);
+	irq = irq & 1 << INTR_IIR;
+	pci_write_config_dword(dev, PCI_INTERRUPT_CTRL, irq);
+
+	pci_read_config_dword(dev, PCI_INTERRUPT_CTRL, &irq);
+	irq &= ~(1 << INTR_IER);
+	pci_write_config_dword(dev, PCI_INTERRUPT_CTRL, irq);
+
+	atomisp_msi_irq_uninit(isp, dev);
+
+	atomisp_ospm_dphy_down(isp);
+	if (ATOMISP_INTERNAL_PM) {
+		if (atomisp_mrfld_power_down(isp))
+			dev_err(&dev->dev, "Failed to switch off ISP\n");
+	}
+
 	pci_dev_put(isp->pci_root);
 	return err;
 }
-- 
1.9.3

